File: internal/services/git.go
package services

import (
	"fmt"
	"os"
	"os/exec"
	"strings"
)

// ── GitRunner interface implementation ──

func (r *ExecGitRunner) ListWorktrees(dir string) []GitWorktree {
	out, err := exec.Command("git", "-C", dir, "worktree", "list", "--porcelain").Output()
	if err != nil {
		return nil
	}

	var result []GitWorktree
	var currentPath, currentBranch string
	for _, line := range strings.Split(string(out), "\n") {
		if path, ok := strings.CutPrefix(line, "worktree "); ok {
			currentPath = path
		} else if branch, ok := strings.CutPrefix(line, "branch refs/heads/"); ok {
			currentBranch = branch
		} else if line == "" && currentPath != "" {
			if currentBranch != "" {
				result = append(result, GitWorktree{Path: currentPath, Branch: currentBranch})
			}
			currentPath = ""
			currentBranch = ""
		}
	}
	if currentPath != "" && currentBranch != "" {
		result = append(result, GitWorktree{Path: currentPath, Branch: currentBranch})
	}
	return result
}

func (r *ExecGitRunner) CurrentBranch(dir string) string {
	out, err := exec.Command("git", "-C", dir, "rev-parse", "--abbrev-ref", "HEAD").Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

func (r *ExecGitRunner) CreateWorktreeFromBase(repoDir, newBranch, baseBranch string, copyFilesList []string, workspaceDir string) (string, error) {
	repoName := fileBase(repoDir)

	var worktreeDir string
	if workspaceDir != "" {
		worktreeDir = workspaceDir + "/" + repoName
	} else {
		dirSuffix := strings.ReplaceAll(newBranch, "/", "-")
		worktreeDir = repoDir + "/../" + repoName + "--" + dirSuffix
	}

	DockerForceCleanup(DockerProjectName(worktreeDir))

	if _, err := os.Stat(worktreeDir); err == nil {
		return "", fmt.Errorf("worktree directory already exists: %s", worktreeDir)
	}

	branchExists := exec.Command("git", "-C", repoDir, "rev-parse", "--verify", newBranch).Run() == nil
	remoteExists := !branchExists && exec.Command("git", "-C", repoDir, "rev-parse", "--verify", "origin/"+newBranch).Run() == nil

	var cmd *exec.Cmd
	if branchExists {
		cmd = exec.Command("git", "-C", repoDir, "worktree", "add", worktreeDir, newBranch)
	} else if remoteExists {
		cmd = exec.Command("git", "-C", repoDir, "worktree", "add", "--track", "-b", newBranch, worktreeDir, "origin/"+newBranch)
	} else {
		cmd = exec.Command("git", "-C", repoDir, "worktree", "add", "-b", newBranch, worktreeDir, baseBranch)
	}

	if out, err := cmd.CombinedOutput(); err != nil {
		return "", fmt.Errorf("git worktree add: %s (%w)", strings.TrimSpace(string(out)), err)
	}

	CopyFiles(repoDir, worktreeDir, copyFilesList)
	return worktreeDir, nil
}

func (r *ExecGitRunner) RemoveWorktree(repoDir, worktreePath, branch string) error {
	if info, err := os.Stat(worktreePath); err == nil && info.IsDir() {
		if _, err := os.Stat(worktreePath + "/docker-compose.yml"); err == nil {
			cmd := exec.Command("docker", "compose", "down", "-v", "--remove-orphans", "--timeout", "5")
			cmd.Dir = worktreePath
			_ = cmd.Run()
		}
	}

	DockerForceCleanup(DockerProjectName(worktreePath))
	_ = exec.Command("git", "-C", repoDir, "worktree", "remove", "--force", worktreePath).Run()

	if _, err := os.Stat(worktreePath); err == nil {
		_ = os.RemoveAll(worktreePath)
	}

	_ = exec.Command("git", "-C", repoDir, "worktree", "prune").Run()
	_ = exec.Command("git", "-C", repoDir, "branch", "-D", branch).Run()
	return nil
}

// ── Package-level functions (delegate to DefaultGit) ──

func ListWorktrees(dir string) []GitWorktree { return DefaultGit.ListWorktrees(dir) }
func CurrentBranch(dir string) string         { return DefaultGit.CurrentBranch(dir) }

func CreateWorktreeFromBase(repoDir, newBranch, baseBranch string, copyFiles []string, workspaceDir string) (string, error) {
	return DefaultGit.CreateWorktreeFromBase(repoDir, newBranch, baseBranch, copyFiles, workspaceDir)
}

func RemoveWorktree(repoDir, worktreePath, branch string) error {
	return DefaultGit.RemoveWorktree(repoDir, worktreePath, branch)
}

// ── Non-interface git helpers ──

func IsBranchInWorktree(dir, branch string) bool {
	for _, wt := range ListWorktrees(dir) {
		if wt.Branch == branch {
			return true
		}
	}
	return false
}

func Checkout(dir, branch string) error {
	out, err := exec.Command("git", "-C", dir, "checkout", branch).CombinedOutput()
	if err != nil {
		return fmt.Errorf("git checkout %s: %s (%w)", branch, strings.TrimSpace(string(out)), err)
	}
	return nil
}

func CheckoutNewBranch(dir, branch string) error {
	out, err := exec.Command("git", "-C", dir, "checkout", "-b", branch).CombinedOutput()
	if err != nil {
		return fmt.Errorf("git checkout -b %s: %s (%w)", branch, strings.TrimSpace(string(out)), err)
	}
	return nil
}

func CreateWorktree(repoDir, branch string, copyFilesList []string) (string, error) {
	dirSuffix := strings.ReplaceAll(branch, "/", "-")
	repoName := fileBase(repoDir)
	parent := repoDir + "/.."
	worktreeDir := parent + "/" + repoName + "--" + dirSuffix

	if _, err := os.Stat(worktreeDir); err == nil {
		return "", fmt.Errorf("worktree directory already exists: %s", worktreeDir)
	}

	out, err := exec.Command("git", "-C", repoDir, "worktree", "add", worktreeDir, branch).CombinedOutput()
	if err != nil {
		stderr := string(out)
		if strings.Contains(stderr, "already checked out") || strings.Contains(stderr, "is already used") {
			newBranch := "wt/" + dirSuffix
			out2, err2 := exec.Command("git", "-C", repoDir, "worktree", "add", worktreeDir, newBranch).CombinedOutput()
			if err2 != nil {
				out3, err3 := exec.Command("git", "-C", repoDir, "worktree", "add", "-b", newBranch, worktreeDir, branch).CombinedOutput()
				if err3 != nil {
					return "", fmt.Errorf("git worktree add failed: %s", strings.TrimSpace(string(out3)))
				}
				_ = out2
			}
		} else {
			return "", fmt.Errorf("git worktree add failed: %s", strings.TrimSpace(stderr))
		}
	}

	CopyFiles(repoDir, worktreeDir, copyFilesList)
	return worktreeDir, nil
}

func fileBase(path string) string {
	parts := strings.Split(strings.TrimRight(path, "/"), "/")
	if len(parts) == 0 {
		return "repo"
	}
	return parts[len(parts)-1]
}

File: src/services/git.rs
use anyhow::{bail, Result};
use std::path::{Path, PathBuf};
use std::process::Command;

use super::docker::{docker_force_cleanup, docker_project_name};
use super::files::copy_files;

/// List existing git worktrees for a repo.
pub fn list_worktrees(dir: &Path) -> Result<Vec<(String, String)>> {
    let output = Command::new("git")
        .args(["-C", &dir.to_string_lossy(), "worktree", "list", "--porcelain"])
        .output()?;

    if !output.status.success() {
        return Ok(Vec::new());
    }

    let text = String::from_utf8_lossy(&output.stdout);
    let mut result = Vec::new();
    let mut current_path = String::new();
    let mut current_branch = String::new();

    for line in text.lines() {
        if let Some(path) = line.strip_prefix("worktree ") {
            current_path = path.to_string();
        } else if let Some(branch) = line.strip_prefix("branch refs/heads/") {
            current_branch = branch.to_string();
        } else if line.is_empty() && !current_path.is_empty() {
            if !current_branch.is_empty() {
                result.push((
                    std::mem::take(&mut current_path),
                    std::mem::take(&mut current_branch),
                ));
            }
            current_path.clear();
            current_branch.clear();
        }
    }
    if !current_path.is_empty() && !current_branch.is_empty() {
        result.push((current_path, current_branch));
    }

    Ok(result)
}

/// Check if a branch is already checked out in any worktree for this repo.
pub fn is_branch_in_worktree(dir: &Path, branch: &str) -> bool {
    list_worktrees(dir).unwrap_or_default().iter()
        .any(|(_, b)| b == branch)
}

/// Create a git worktree with a NEW branch from a base branch.
pub fn create_worktree_from_base(
    repo_dir: &Path,
    new_branch: &str,
    base_branch: &str,
    copy_files_list: &[String],
    workspace_dir: Option<&Path>,
) -> Result<PathBuf> {
    let repo_name = repo_dir.file_name()
        .map(|n| n.to_string_lossy().to_string())
        .unwrap_or_else(|| "repo".to_string());

    let worktree_dir = if let Some(ws_dir) = workspace_dir {
        ws_dir.join(&repo_name)
    } else {
        let dir_suffix = new_branch.replace('/', "-");
        let parent = repo_dir.parent().unwrap_or(Path::new("."));
        parent.join(format!("{repo_name}--{dir_suffix}"))
    };

    let project = docker_project_name(&worktree_dir);
    docker_force_cleanup(&project);

    if worktree_dir.exists() {
        bail!("worktree directory already exists: {}", worktree_dir.display());
    }

    // Check if branch already exists (local or remote)
    let branch_exists = Command::new("git")
        .args(["-C", &repo_dir.to_string_lossy(), "rev-parse", "--verify", new_branch])
        .output()
        .is_ok_and(|o| o.status.success());

    let remote_exists = !branch_exists && Command::new("git")
        .args(["-C", &repo_dir.to_string_lossy(), "rev-parse", "--verify", &format!("origin/{new_branch}")])
        .output()
        .is_ok_and(|o| o.status.success());

    let output = if branch_exists {
        // Checkout existing local branch
        Command::new("git")
            .args([
                "-C", &repo_dir.to_string_lossy(),
                "worktree", "add",
                &worktree_dir.to_string_lossy(),
                new_branch,
            ])
            .output()?
    } else if remote_exists {
        // Checkout existing remote branch
        Command::new("git")
            .args([
                "-C", &repo_dir.to_string_lossy(),
                "worktree", "add",
                "--track", "-b", new_branch,
                &worktree_dir.to_string_lossy(),
                &format!("origin/{new_branch}"),
            ])
            .output()?
    } else {
        // Create new branch from base
        Command::new("git")
            .args([
                "-C", &repo_dir.to_string_lossy(),
                "worktree", "add",
                "-b", new_branch,
                &worktree_dir.to_string_lossy(),
                base_branch,
            ])
            .output()?
    };

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        bail!("git worktree add failed: {stderr}");
    }

    copy_files(repo_dir, &worktree_dir, copy_files_list);
    Ok(worktree_dir)
}

/// Create a git worktree checking out an EXISTING branch.
pub fn create_worktree(repo_dir: &Path, branch: &str, copy_files_list: &[String]) -> Result<PathBuf> {
    let dir_suffix = branch.replace('/', "-");
    let parent = repo_dir.parent().unwrap_or(Path::new("."));
    let repo_name = repo_dir.file_name()
        .map(|n| n.to_string_lossy().to_string())
        .unwrap_or_else(|| "repo".to_string());
    let worktree_dir = parent.join(format!("{repo_name}--{dir_suffix}"));

    if worktree_dir.exists() {
        bail!("worktree directory already exists: {}", worktree_dir.display());
    }

    let output = Command::new("git")
        .args([
            "-C", &repo_dir.to_string_lossy(),
            "worktree", "add",
            &worktree_dir.to_string_lossy(),
            branch,
        ])
        .output()?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        if stderr.contains("already checked out") || stderr.contains("is already used") {
            let new_branch = format!("wt/{}", dir_suffix);
            let output2 = Command::new("git")
                .args([
                    "-C", &repo_dir.to_string_lossy(),
                    "worktree", "add",
                    &worktree_dir.to_string_lossy(),
                    &new_branch,
                ])
                .output()?;
            if !output2.status.success() {
                let output3 = Command::new("git")
                    .args([
                        "-C", &repo_dir.to_string_lossy(),
                        "worktree", "add",
                        "-b", &new_branch,
                        &worktree_dir.to_string_lossy(),
                        branch,
                    ])
                    .output()?;
                if !output3.status.success() {
                    let stderr3 = String::from_utf8_lossy(&output3.stderr);
                    bail!("git worktree add failed: {stderr3}");
                }
            }
        } else {
            bail!("git worktree add failed: {stderr}");
        }
    }

    copy_files(repo_dir, &worktree_dir, copy_files_list);
    Ok(worktree_dir)
}

/// Remove a git worktree and clean up its branch + leftover files.
pub fn remove_worktree(repo_dir: &Path, worktree_path: &Path, branch: &str) -> Result<()> {
    if worktree_path.exists() && worktree_path.join("docker-compose.yml").is_file() {
        let _ = Command::new("docker")
            .args(["compose", "down", "-v", "--remove-orphans", "--timeout", "5"])
            .current_dir(worktree_path)
            .output();
    }

    let project = docker_project_name(worktree_path);
    docker_force_cleanup(&project);

    let _ = Command::new("git")
        .args([
            "-C", &repo_dir.to_string_lossy(),
            "worktree", "remove", "--force",
            &worktree_path.to_string_lossy(),
        ])
        .output();

    if worktree_path.exists() {
        let _ = std::fs::remove_dir_all(worktree_path);
    }

    let _ = Command::new("git")
        .args(["-C", &repo_dir.to_string_lossy(), "worktree", "prune"])
        .output();

    let _ = Command::new("git")
        .args(["-C", &repo_dir.to_string_lossy(), "branch", "-D", branch])
        .output();

    Ok(())
}

// ── Branch operations ──

pub fn current_branch(dir: &str) -> Option<String> {
    let output = Command::new("git")
        .args(["-C", dir, "rev-parse", "--abbrev-ref", "HEAD"])
        .output()
        .ok()?;
    if !output.status.success() { return None; }
    let branch = String::from_utf8_lossy(&output.stdout).trim().to_string();
    if branch.is_empty() { None } else { Some(branch) }
}

pub fn checkout(dir: &str, branch: &str) -> Result<()> {
    let output = Command::new("git")
        .args(["-C", dir, "checkout", branch])
        .output()?;
    if !output.status.success() {
        bail!("{}", String::from_utf8_lossy(&output.stderr).trim());
    }
    Ok(())
}

pub fn checkout_new_branch(dir: &str, branch: &str) -> Result<()> {
    let output = Command::new("git")
        .args(["-C", dir, "checkout", "-b", branch])
        .output()?;
    if !output.status.success() {
        bail!("{}", String::from_utf8_lossy(&output.stderr).trim());
    }
    Ok(())
}
